Ortto Action Destination

xitonix opened the pull request:
<!-- Hello and thank you for contributing to Segment action-destinations! -->

<!-- Before opening your pull request, make sure you have added and ran unit
     tests and tested your change locally. Refer to our testing
     documentation for more information: https://github.com/segmentio/action-destinations/blob/main/docs/testing.md -->

<!-- If you have questions or issues please open a new issue or create a new discussion
     post in Github. -->

_A summary of your pull request, including the what change you're making and why._

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [If destination is already live] Tested for backward compatibility of destination. **Note:** New required fields are a breaking change.
- [ ] [Segmenters] Tested in the staging environment
- [ ] [Segmenters] [If applicable for this change] Tested for regression with Hadron.

joe-ayoub-segment commented:
hi @xitonix thanks for raising this PR for a new Integration. Apologies for the delay reviewing this - we've had some staffing availability issues lately. 

I'm going to review this in the next 24h.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/ortto/common-fields.ts:
```diff
@@ -0,0 +1,144 @@
+    description: `Ortto audience ID`,
+    type: 'string',
+    default: {
+      '@if': {
+        exists: {
+          '@path': '$.context.personas.audience_settings.audience_id'
+        },
+        then: { '@path': '$.context.personas.audience_settings.audience_id' },
+        else: { '@path': '$.context.personas.external_audience_id' }
+      }
```
Note to self: I need to check that these default mappings make sense. 
@xitonix - why do you want to check the location '$.context.personas.audience_settings.audience_id' ?
The Audience ID should always be located at '$.context.personas.external_audience_id'

joe-ayoub-segment replied:
OK cool - we can focus on this when we meet in person next week.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/ortto/ortto-client.ts:
```diff
@@ -0,0 +1,209 @@
+    if (!audienceName) {
+      throw new PayloadValidationError(Errors.MissingAudienceName)
+    }
+    const url = this.getEndpoint(settings.api_key).concat('/audiences/create')
```
Is the call to this API idempotent?
i.e what happens if the Audience has already been created? Will it simply return the pre-existing Audience details? Only reason I ask is if the customer creates an Audience, then deletes it, then recreates it with the same name.

xitonix replied:
That's correct — if an audience with the same name already exists in our system, it will be returned.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/ortto/utils.ts:
```diff
@@ -0,0 +1,8 @@
+export const TEST_API_KEY = 'pau-key'
+export function cleanObject<T extends {}>(obj: T): Partial<T> {
+  return Object.fromEntries(
+    Object.entries(obj)
+      .filter(([key, _]) => key !== '')
+      .map(([key, value]) => [key, value instanceof Object && !(value instanceof Array) ? cleanObject(value) : value])
+  ) as Partial<T>
+}
```
Would something like this work better? It doesn't use recursion: 

```
function cleanObject<T>(obj: T): Partial<T> {
  return JSON.parse(
    JSON.stringify(obj, (key, value) =>
      key === '' ? undefined : value
    )
  );
}
```

xitonix replied:
Doesn't seem like it. Looks like `parse` doesn't like `undefined`
![2025-05-02 at 09 27 08](https://github.com/user-attachments/assets/69348ae3-d17d-441e-9d12-01406672746c)

joe-ayoub-segment replied:
Let's leave as is then.

joe-ayoub-segment commented on packages/destination-actions/src/destinations/ortto/ortto-client.ts:
```diff
@@ -0,0 +1,209 @@
+      filtered.push(cleanObject(event))
+    }
+    if (filtered.length == 0) {
+      return
```
Maybe throw an error if there are no valid payloads?

xitonix replied:
I don’t think we should. If an individual payload is truly invalid, we already return a `PayloadValidationError`. In this case, it’s reasonable to silently drop track events sent by Ortto— as users can send activities from Ortto to Segment, and silent dropping will prevent infinite flow of data. It's a valid scenario imo, not an error state.

joe-ayoub-segment reviewed (changes requested):
hi @xitonix thanks for this substantial PR for a new Integration! 

I think the next step should be for you to look through the comments I left, and then for us to catch up 1:1. 

I have some questions about the design of the Integration. 

In particular how Contacts are added and removed from Audiences. 

If I recall correctly you are based in Australia? I'm in the UK timezone but can meet at a time that works for you between 10am UK time and 9pm UK time. Please suggest a time and I'll forward you an invite. I can meet tomorrow (Friday), Monday or Tuesday. 

I look forward to catching up. 
Best regards,
Joe

xitonix commented on packages/destination-actions/src/destinations/ortto/common-fields.ts:
```diff
@@ -0,0 +1,144 @@
+      },
+      phone: {
+        label: 'Phone Number',
+        description: "The contact's phone number",
```
Country code isn’t mandatory in our system, as we fall back to an account-level default when it's not provided—and our customers are aware of this. That said, it's still highly recommended, and I’ll update the description to reflect that.

xitonix commented on packages/destination-actions/src/destinations/ortto/ortto-client.ts:
```diff
@@ -0,0 +1,209 @@
+    })
+  }
+
+  sendActivities = async (settings: Settings, payloads: EventPayload[], hookAudienceID: string) => {
```
Certainly, mate. At the moment, our endpoints mimic that behavior by returning something like this:
```
...
received: 100,
errors: {
  "error message": [even_id_1,...., event_id_N]
}
```
The HTTP status is determined based on the number of errors:

- 400 if all events failed,
- 202 if everything succeeded,
- 206 if it’s a partial success (some valid, some invalid).

I’m happy to revisit the logic if there’s a better approach you’d recommend.

joe-ayoub-segment replied:
Thanks for explaining. 
It would be ideal if an array of response codes were provided to indicate which requests were successful or not. But we could still partially implement multistatus responses for cases where some payloads fail validation (which prevents a request from being sent to your platform). 
Let's chat about this when we meet next week.

xitonix commented on packages/destination-actions/src/destinations/ortto/trackActivity/index.ts:
```diff
@@ -0,0 +1,204 @@
+      default: { '@path': '$.context.ip' },
+      allowNull: true
+    },
+    location: {
+      label: 'Location',
+      description: 'The location where the activity occurred. Will take priority over the IP address.',
+      type: 'object',
+      defaultObjectUI: 'keyvalue:only',
+      additionalProperties: false,
+      allowNull: true,
+      properties: {
+        country: {
+          label: 'Country',
+          type: 'string',
+          allowNull: true
+        },
+        state: {
+          label: 'State',
+          type: 'string',
+          allowNull: true
+        },
+        city: {
+          label: 'City',
+          type: 'string',
+          allowNull: true
+        },
+        post_code: {
+          label: 'Postcode',
+          type: 'string',
+          allowNull: true
+        }
+      },
```
Good catch! I likely forgot to use the common fields for this action. I’ve now also extracted the hook input/output fields into the same file to keep things consistent. Hopefully they work as expected—though I’m not sure how to test them locally since the Action Tester doesn’t display those fields.

xitonix commented on packages/destination-actions/src/destinations/ortto/trackActivity/index.ts:
```diff
@@ -0,0 +1,204 @@
+      },
+      outputTypes: {
+        id: {
+          type: 'string',
+          label: 'ID',
+          description: 'The ID of the Ortto audience to which the associated contacts will be synced.',
+          required: false
+        },
+        name: {
+          type: 'string',
+          label: 'Name',
+          description: 'The name of the Ortto audience to which the associated contacts will be synced.',
+          required: false
+        }
+      },
+      performHook: async (request, { settings, hookInputs }) => {
+        if (hookInputs.name) {
+          try {
+            const client: OrttoClient = new OrttoClient(request)
+            const audience = await client.createAudience(settings, hookInputs.name as string)
+            return {
+              successMessage: `Audience '${audience.name}' (id: ${audience.id}) has been created successfully.`,
+              savedData: {
+                id: audience.id,
+                name: audience.name
+              }
+            }
+          } catch (err) {
+            return {
+              error: {
+                message: (err as APIError).message ?? 'Unknown Error',
+                code: (err as APIError).status?.toString() ?? 'Unknown Error'
+              }
+            }
+          }
+        }
+        return {}
+      }
+    }
+  },
```
That's correct, @joe-ayoub-segment. The goal is to allow users to optionally sync contacts with an Ortto Audience, even if they don’t use the Audience feature in Segment.

> I know it’s been a while since we last met, so my memory might be a little rusty (apologies!).

No worries, mate. When we last spoke, we didn’t have plans to support Audiences, but it's been a feature requested by our mutual customers.

> And you're looking to ensure that the Audience is created on your platform before any user-tracked activity data is sent to it. Is that correct?

Yes, that's right, **only if the user chooses to**. Otherwise, they can simply use an existing audience by providing the `audience_id` field. Both options are optional.

> If so, are you expecting customers to fill in both the audience_id field and the inputFields.name field with the same value?

No, that’s not the case. If the user doesn’t have an existing audience and wants to create one when saving the mapping, they can do so by providing a `name` via `retlOnMappingSave`. If the `audience_id` field is set and a new audience is created on save, the existing audience selected by `audience_id` will take precedence.

At least that’s how I understand the behaviour of `retlOnMappingSave`. Please let me know if I’m on the wrong track!

joe-ayoub-segment replied:
Let's also chat about this when we meet!

xitonix commented:
Hi @joe-ayoub-segment,

Thank you so much for the detailed review — I really appreciate the time you took to go through everything. I’ve gone ahead and addressed/responded to your comments.

Regarding the 1:1, mornings on your end would work best for me as your 9 AM is 6 PM here in Sydney. Unfortunately, I won’t be able to make it this evening, but I’m available either Monday or Tuesday morning UK time. Please feel free to schedule a meeting for whichever of those days suits you best.

Thanks again for the review — looking forward to catching up!

Best regards,
Alex

joe-ayoub-segment commented on packages/destination-actions/src/destinations/ortto/index.ts:
```diff
@@ -0,0 +1,98 @@
+      headers: {
+        Authorization: `Bearer ${settings.api_key}`
+      },
+      timeout: Math.max(30_000, DEFAULT_REQUEST_TIMEOUT)
```
Thanks @xitonix I think it would be a good idea to not use the timeout for the moment. If needed we can re-apply it.

joe-ayoub-segment commented:
Thanks @xitonix - I've scheduled a call at 9am my time next Monday. Feel free to add any other attendees.

joe-ayoub-segment commented:
hi @xitonix - would you be able to schedule a call with me using this Calendly link please?

https://calendly.com/joe_ayoub/integration-workshop

I look forward to catching up, 
Joe

xitonix commented:
Hey @joe-ayoub-segment,
As discussed, I’ve removed the Engage-related actions and added support for multi-status responses.
I’ll add Audience support in a separate PR/integration.
Could you please take another look when you have a moment? Thanks! 🙏🏻
